readme.md



## SQLite3 used for development with Prisma ORM

#### Schema
```
/schema.prisma

datasource db {
    url = "file:./app.db"
    provider = "sqlite"
}

generator client {
    provider = "prisma-client-js"
}


model User {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  email     String   @unique
  username String @unique
  name      String?
  password  String   @unique
  resetPass String?  @unique
  resetExp  Int? 
}
```
#### Run migration
```
npm run migrate:dev
```
---
## Endpoints
### User dashboard with valid webtoken
```
app.get('/dashboard', verify, (req: express.Request, res: express.Response) => {

    return res.json({
        title: 'Dashboard',
        page: 'user-dashboard'
    })

})
```

### User login
```app.post('/login', async (req: express.Request, res: express.Response) => {

    try {
        const user = await prisma.user.findUnique({
            where: { username: req.body.username },
            select: {
                password: true,
                retry: true,
                retryExp: true
            }
        })
        const pass = await comparePw(user?.password, req.body.password)


        if (pass && Number(user?.retry) <= 3) {
            const signedToken = await sign()

            await prisma.user.update({
                where: {
                    email: req.body.email
                },
                data: {
                    retry: 0,
                    retryExp: 0
                }
            })


            await res.cookie('accessToken', signedToken, { httpOnly: true, expires: new Date(Date.now() + 8 * 3600000), path: '/dashboard' })
            return res.json({
                msg: pass
            })
        }
        else {
            if (user) {


                switch (user.retry) {

                    case 3:
                        if(Number(user.retryExp) > new Date(Date.now()).valueOf()){

                            res.json({
                                message: "Your account is still locked. Please contact support to unlock your account",
                                now: new Date(Date.now()).valueOf(),
                                exp: Number(user.retryExp)
                            })

                        }
                        else{

                            await prisma.user.update({
                                where: {
                                    email: req.body.email
                                },
                                data: {
                                    retry: 1,
                                }
                            })
                            res.json({
                                message: "Your account is unlocked",
                                now: new Date(Date.now()).valueOf(),
                                exp: Number(user.retryExp)
                            })

                        }
                        break;
                    case 2:
                        await prisma.user.update({
                            where: {
                                email: req.body.email
                            },
                            data: {
                                retry:{
                                    increment: 1
                                },
                                retryExp: new Date(Date.now() + RETRY_TIMER * 60000).valueOf()
                            }
                        })
                        res.json({
                            message: `Your account has been locked for ${RETRY_TIMER} minute(s). Please try again later`
                        })
                        break;

                    default:
                        await prisma.user.update({
                            where: {
                                email: req.body.email
                            },
                            data: {
                                retry: {
                                    increment: 1
                                }
                            }
                        })
                        res.json({
                            message: "Something went wrong, please try again. "
                        })


                }




            }
            else {

                res.redirect(302, '/login')
            }

        }
    }
    catch (err) {
        console.log(err)
        res.status(500).json({
            message: "Something has gone wrong please try again"
        })
    }

})``

```


### User registration 
```
app.post('/register', async (req: express.Request, res: express.Response) => {

    try {
        const hashed: any = await hashPw(req.body.password)
        const newUser = await prisma.user.create({
            data: {
                name: req.body.name,
                username: req.body.username,
                email: req.body.email,
                password: hashed
            },
        })
        return res.json({
            message: `${req.body.username} account has been created`
        })
    }
    catch (error) {
        return res.json({
            error: error
        })
    }

})
```

### User forgot password
```
app.post('/forgot-password', async (req: express.Request, res: express.Response) => {
    try {
        const getUser = await prisma.user.findUnique({
            where: {
                email: req.body.email
            }
        })

        if (getUser != null) {
            const token = await genHash(RESET_BYTES)
            const addResetToken = await prisma.user.update({
                where: {
                    email: req.body.email
                },
                data: {
                    resetPass: token,
                    resetExp: new Date(Date.now() + MINUTES * 60000).valueOf()
                }
            })
            return res.json({
                resetPasswordToken: token,
                resetPasswordExpiration: new Date(Date.now() + MINUTES * 60000).valueOf(),
                tokenExpiresIn: `${MINUTES} minutes`,
                urlSentToEmail: `${SITE}/reset-password/${token.toString()}`
            })
        }
        else {
            return res.status(400).json({
                message: "Email doesn't exist"
            })
        }


    }
    catch (e) {
        return res.status(500).json({
            msg: e
        })
    }



})
```

### User resets password with valid token
```
app.post('/reset-password/:token', async (req: express.Request, res: express.Response) => {

    try {

        const user = await prisma.user.findUnique({
            where: {
                resetPass: req.params.token
            }
        })
        const now = new Date(Date.now()).valueOf()

        if (user?.resetExp) {
            if (user.resetExp > now) {
                const hashed: any = await hashPw(req.body.password)
                const addResetToken = await prisma.user.update({
                    where: {
                        email: user.email
                    },
                    data: {
                        password: hashed,
                        resetExp: 0,
                        resetPass: ""
                    }
                })

                return res.json({
                    expiration: user.resetExp,
                    now: now,
                    expiredToken: user.resetExp > now ? true : false,
                    message: `${user.username} account password has been reset`

                })
            }
            else {

                const addResetToken = await prisma.user.update({
                    where: {
                        email: user.email
                    },
                    data: {
                        resetExp: 0,
                        resetPass: ""
                    }
                })

                return res.json({
                    message: "Reset password has expired. Please try again if needed.",
                    expiredToken: user.resetExp > now ? true : false,
                    expiration: user.resetExp,
                    now: now
                })
            }

        }
        else {
            return res.json({
                message: `Are you trying to reset your password? Visit ${SITE}/forgot-password to reset your password`,
            })
        }


    }
    catch (e) {
        return res.status(500).json({
            message: "Something went wrong please try again"
        })
    }


})

```

### User resets password with valid token
```
app.get('/forbidden', (req: express.Request, res: express.Response) => {

    return res.json({
        msg: "Please try again"
    })
})
```
---
## Middleware and methods

### Verify if token is valid
```
export const verify = async (req: express.Request, res: express.Response, next: express.NextFunction) => {
 
    try {
        if (!await jwt.verify(req.cookies.accessToken, "process.env.jwtSecret")) res.redirect("403")
            next()
    }
    catch (err) {
        res.redirect(403, '/login')
    }
}
```

### Sign Token
```
export const sign = async () => {
    const signed = await jwt.sign({
        type: 'appUser',
        role: 'appPaid'
    }, "process.env.jwtSecret",{
        expiresIn: '30s'
    })
    return signed
}
```


### Hash password

```
export const hashPw = async (pw: string) => {

    try {
        const hash = await ag.hash(pw)
        return hash
    }
    catch (err) {
        return err
    }
}

```

### Basic hash generator

```
export const genHash = (bytes: number) => {
    return crypto.randomBytes(bytes).toString('hex')
}
```

### Verify password
```
export const comparePw = async (hash: any, pw: string) => {
  
    if (await ag.verify(hash, pw)) {
        return true
    }
    else {
        return false
    }
}

```

changelog.md
# Changelog
Changes to this project will be documented in this file

## 2022-03-22
### Change
- change code base to typescript
- change from mongodb to sqlite3 for development
### Added 
- Forgot/Reset password pattern
- Added Prisma ORM
- Added Login retry logic - locks account after 3 attempts

src/fn.js
const argon2 = require('argon2')
const jwt = require('jsonwebtoken')
const msg = require('./msg')
const dotenv = require('dotenv')
dotenv.config()


module.exports.verify = async (req, res, next) => {
 
    try {
        if (!await jwt.verify(req.cookies.accessToken, process.env.jwtSecret)) res.redirect(403)
            next()
    }
    catch (err) {
        res.redirect(403, '/login')
    }
}

module.exports.sign = async () => {
    const signed = await jwt.sign({
        type: 'appUser'
    }, process.env.jwtSecret)
    return signed
}

module.exports.hashPw = async (pw) => {

    try {
        const hash = await argon2.hash(pw)
        return hash
    }
    catch (err) {
        return err
    }

}

module.exports.comparePw = async (hash, pw) => {
  
    if (await argon2.verify(hash, pw)) {
        return true
    }
    else {
        return false
    }
}



src/fn.ts

import * as ag from 'argon2'
import * as jwt from 'jsonwebtoken'
import * as msg from './msg'
import * as dotenv from 'dotenv'
import * as express from 'express'
import * as crypto from 'crypto'

dotenv.config()

export const verify = async (req: express.Request, res: express.Response, next: express.NextFunction) => {
 
    try {
        if (!await jwt.verify(req.cookies.accessToken, "process.env.jwtSecret")) res.redirect("403")
            next()
    }
    catch (err) {
        res.redirect(403, '/login')
    }
}

export const sign = async () => {
    const signed = await jwt.sign({
        type: 'appUser',
        role: 'appPaid'
    }, "process.env.jwtSecret",{
        expiresIn: '30s'
    })
    return signed
}

export const hashPw = async (pw: string) => {

    try {
        const hash = await ag.hash(pw)
        return hash
    }
    catch (err) {
        return err
    }

}

export const resetpass = async () => {
    const signed = await jwt.sign({
        type: 'appUser',
        action: 'reset'
    }, "process.env.jwtSecret",{
        expiresIn: '5m'
    })
    return signed
}

export const genHash = (bytes: number) => {
    return crypto.randomBytes(bytes).toString('hex')
}

export const comparePw = async (hash: any, pw: string) => {
  
    if (await ag.verify(hash, pw)) {
        return true
    }
    else {
        return false
    }
}



src/app.js


const Express = require('express')
const app = Express()
const mg = require('mongoose')

const bp = require('body-parser')
const cors = require('cors')
const fn = require('./fn')
const msg = require('./msg')
const dotenv = require('dotenv')
const cookiPar = require('cookie-parser')

dotenv.config()

/**
 * Models
 */
const user = require('./models/User')
const { redirect } = require('express/lib/response')
const cookieParser = require('cookie-parser')

const PORT = 3000
const HOST = '0.0.0.0'

app.use(bp.json())
app.use(bp.urlencoded({extended:true}))
app.use(cookieParser())
app.use(cors())

const dbUser = 'admin'
const dbPw = 'adminpwd'
const dbHost = 'localhost'
const dbPort = '27018'
const dbName = 'authDb'

const connect = async () => {
    await mg.connect(`mongodb://${dbUser}:${dbPw}@${dbHost}:${dbPort}/${dbName}?authSource=admin`)
  
}
connect().catch(error => console.log("Connection Error: "+ error))


/**
 * path: /
 */
app.get('/',(req,res) => {
    
    return res.json({
        "title":"Jwt Authentication Service",
        "author":"bshelling@gmail.com",
        "version":"1.0"
    })

}) 


/**
 * path: /paywall
 */
app.get('/dashboard',fn.verify,(req,res) => {
    
    
    return res.json({
        title: 'Dashboard',
        page: 'user-dashboard'
    })
     
})


/**
 * path: /authenticate
 */
app.post('/authenticate',async (req,res)=>{

    try {
    
        const usr = await user.User.findOne({username:req.body.username}).exec()
        if(!await fn.comparePw(usr.password,req.body.password)) return redirect(403,'/forbidden')
        const signedToken = await fn.sign()
        await res.cookie('accessToken',signedToken, {httpOnly: true} )
        return res.json({
            msg: await fn.comparePw(usr.password,req.body.password)
        })


    }
    catch(err){
        
        res.redirect(403,'/authenticate')
    }

})

/**
 * path: /register
 */
app.post('/register',async (req,res) =>{

    try{
        if(await user.User.exists({username : req.body.username})) return res.json({msg:msg.userExists});
        const usr = new user.User({
            username: req.body.username,
            password: await fn.hashPw(req.body.password)
        })
        await usr.save()
        return res.json({
            message: `${req.body.username} account has been created`
        })
    }
    catch(error){
        return res.json({
            error: error
        })
    }

})

/**
 * path: /reset-password
 */
app.post('/reset-password',(req,res) =>{


    return res.json({
        username: req.body.username,
        password: req.body.password
    })


})

app.get('/forbidden',(req,res)=>{

    return res.json({
        msg: "Please try again"
    })
})

app.listen(PORT,HOST,(err) => {
    console.log(`http://localhost:${PORT}`)
})

src/app.ts
/**
 * Import dependencies
 */
import express from 'express'
import * as bp from 'body-parser'
import * as cors from 'cors'
import * as cookieParser from 'cookie-parser'
import * as dotenv from 'dotenv'
import { verify, comparePw, hashPw, sign, resetpass, genHash } from './fn'
import * as msg from './msg'
import { PrismaClient } from '@prisma/client'

/**
 * Initialize dotenv for environment variables
 */
dotenv.config()

const app = express()
app.use(bp.json())
app.use(bp.urlencoded({ extended: true }))
app.use(cookieParser.default())
app.use(cors.default())

/**
 * Instantiate Prisma Client
 */
const prisma = new PrismaClient()

/**
 * Environment Variables
 */
const TESTPW = ""
const MINUTES = 3
const SITE = 'http://localhost:3000'
const RETRY_BYTES = 16
const RETRY_TIMER = 5
const RESET_BYTES = 20
const PORT = 3000
const HOST = '0.0.0.0'


/**
* path: /
*/
app.get('/', (req: express.Request, res: express.Response) => {

    return res.json([{
        "title": "Jwt Authentication Service",
        "author": "bshelling@gmail.com",
        "version": "1.0"
    }])

})

/**
* path: /Dashboard
* - User access dashboard
* - Stored jwt is sent with request through verify middleware.
* - If jwt is valid user access is granted if not user is redirected to /forbidden
*/
app.get('/dashboard', verify, (req: express.Request, res: express.Response) => {

    // Added for development feedback
    return res.json({
        title: 'Dashboard',
        page: 'user-dashboard'
    })

})

/**
* path: /login
* - User enters username and password
* - Username is passed to a database query to retrieve the user
* - Password is passed to the compare plain text with hashed
* - If the password is valid and login retries are < 3 retry and retryExp set to 0, jwt is signed and set to cookie, with an expiration of 8 hours. Then redirect to /dashboard
* - If 3 password retries, account is locked for time period set by RETRY_TIMER variable, after time has expired the user can attempt to login
* - If the user doesn't exist - 400 error with message "Account doesn't exist"
*/
app.post('/login', async (req: express.Request, res: express.Response) => {

    try {
        const user = await prisma.user.findUnique({
            where: { username: req.body.username },
            select: {
                password: true,
                retry: true,
                retryExp: true
            }
        })
        const pass = await comparePw(user?.password, req.body.password)


        if (pass && Number(user?.retry) <= 3) {
            const signedToken = await sign()

            await prisma.user.update({
                where: {
                    email: req.body.email
                },
                data: {
                    retry: 0,
                    retryExp: 0
                }
            })


            await res.cookie('accessToken', signedToken, { httpOnly: true, expires: new Date(Date.now() + 8 * 3600000), path: '/dashboard' })
            return res.redirect('/dashboard')
        }
        else {
            if (user) {

                switch (user.retry) {

                    case 3:
                        if(Number(user.retryExp) > new Date(Date.now()).valueOf()){

                            res.json({
                                message: "Your account is still locked. Please contact support to unlock your account",
                                now: new Date(Date.now()).valueOf(),
                                exp: Number(user.retryExp)
                            })

                        }
                        else{

                            await prisma.user.update({
                                where: {
                                    email: req.body.email
                                },
                                data: {
                                    retry: 1,
                                }
                            })
                            res.json({
                                message: "Your account is unlocked",
                                now: new Date(Date.now()).valueOf(),
                                exp: Number(user.retryExp)
                            })

                        }
                        break;
                    case 2:
                        await prisma.user.update({
                            where: {
                                email: req.body.email
                            },
                            data: {
                                retry:{
                                    increment: 1
                                },
                                retryExp: new Date(Date.now() + RETRY_TIMER * 60000).valueOf()
                            }
                        })
                        res.json({
                            message: `Your account has been locked for ${RETRY_TIMER} minute(s). Please try again later`
                        })
                        break;

                    default:
                        await prisma.user.update({
                            where: {
                                email: req.body.email
                            },
                            data: {
                                retry: {
                                    increment: 1
                                }
                            }
                        })
                        res.json({
                            message: "Something went wrong, please try again. "
                        })


                }




            }
            else {

                res.redirect(302, '/login')
            }

        }
    }
    catch (err) {
        console.log(err)
        res.status(500).json({
            message: "Something has gone wrong please try again"
        })
    }

})

/**
 * path: /register
 * - New user enters username, email, and password
 * - Plain text password is hashed and new user record is stored with username/email/password
 * - Response: message: `${req.body.username} account has been created`
 */
app.post('/register', async (req: express.Request, res: express.Response) => {

    try {
        const hashed: any = await hashPw(req.body.password)
        const newUser = await prisma.user.create({
            data: {
                name: req.body.name,
                username: req.body.username,
                email: req.body.email,
                password: hashed
            },
        })
        return res.json({
            message: `${req.body.username} account has been created`
        })
    }
    catch (error) {
        return res.json({
            error: error
        })
    }

})

/**
 * User requests to reset password
 * path: /reset-password/:token 
 * 
 * - User enters email to request a new password
 * - Email is passed to a database query to retrieve the user
 * - If the user exists - reset password hash is generate and added to the user's record with a token expiration 
 * - If the user doesn't exist - 400 error with message "Account doesn't exist"
 * 
 */
app.post('/forgot-password', async (req: express.Request, res: express.Response) => {
    try {
        const getUser = await prisma.user.findUnique({
            where: {
                email: req.body.email
            }
        })

        if (getUser != null) {
            const token = await genHash(RESET_BYTES)
            const addResetToken = await prisma.user.update({
                where: {
                    email: req.body.email
                },
                data: {
                    resetPass: token,
                    resetExp: new Date(Date.now() + MINUTES * 60000).valueOf()
                }
            })

            // Added for development feedback
            return res.json({
                resetPasswordToken: token,
                resetPasswordExpiration: new Date(Date.now() + MINUTES * 60000).valueOf(),
                tokenExpiresIn: `${MINUTES} minutes`,
                urlSentToEmail: `${SITE}/reset-password/${token.toString()}`
            })
        }
        else {


            return res.status(400).json({
                message: "Account doesn't exist"
            })
        }

    }
    catch (e) {
        return res.status(500).json({
            msg: "Something went wrong please try again"
        })
    }



})


/**
 * User resets password with valid token
 * path: /reset-password/:token 
 * 
 * - User enters a new password and token is sent with request
 * - Reset password token is passed to a database query to retrieve the user
 * - If the reset expiration exists - a check is executed to see if it's expired. 
 * - If the token is not expired, the new password is updated on the record
 * - If the token is expired - response: "Reset password link has expired",
 * - If the token is invalid - response: "Are you trying to reset your password? Visit ${SITE}/forgot-password to reset your password"
 * 
 */
app.post('/reset-password/:token', async (req: express.Request, res: express.Response) => {

    try {

        const user = await prisma.user.findUnique({
            where: {
                resetPass: req.params.token
            }
        })
        const now = new Date(Date.now()).valueOf()

        if (user?.resetExp) {
            if (user.resetExp > now) {
                const hashed: any = await hashPw(req.body.password)
                const addResetToken = await prisma.user.update({
                    where: {
                        email: user.email
                    },
                    data: {
                        password: hashed,
                        resetExp: 0,
                        resetPass: ""
                    }
                })


              // Added for development feedback              
                return res.json({
                    expiration: user.resetExp,
                    now: now,
                    expiredToken: user.resetExp > now ? true : false,
                    message: `${user.username} account password has been reset`

                })
            }
            else {
                
                 // Added for development feedback      
                return res.json({
                    message: "Reset password has expired. Please try again if needed.",
                    expiredToken: user.resetExp > now ? true : false,
                    expiration: user.resetExp,
                    now: now
                })
            }

        }
        else {
            return res.json({
                message: `Are you trying to reset your password? Visit ${SITE}/forgot-password to reset your password`,
            })
        }


    }
    catch (e) {
        return res.status(500).json({
            message: "Something went wrong please try again"
        })
    }


})

/**
 * User logout
 * path: /logout
 * Clears access token when user logs out
 */
app.get('/logout', (req: express.Request, res: express.Response) => {
    res.clearCookie('accessToken',{path:'/dashboard'})
    return res.redirect('/')
})

/**
 * User redirect when token has expired
 * path: /forbidden
 */
app.get('/forbidden', (req: express.Request, res: express.Response) => {

    return res.json({
        msg: "Please try again"
    })
})

app.listen(PORT, HOST, () => {
    console.log(`http://localhost:${PORT}`)
})